File: CrossfitDiary/CrossfitDiary.Web/Scripts/Pages/Models/ViewModels/WorkoutViewModel.ts
module Models {
  import Serializable = General.Serializable;

  interface IWorkoutViewModel {
    id?: number,
    title?: string,
    roundsCount?: number,
    timeToWork?: any,
    timeCap?: any,
    restBetweenExercises?: any,
    restBetweenRounds?: string,
    workoutType: WorkoutType,
    exercisesToDoList: ExerciseViewModel[];
    workoutTypeDisplay?: string;
    children: WorkoutViewModel[];

    isInnerWorkout: boolean;
  }

  export class WorkoutViewModel implements Serializable<WorkoutViewModel>{

    id: number;
    title: string;
    roundsCount?: number;
    timeToWork?: any;
    timeCap?: any;
    restBetweenExercises?: any;
    restBetweenRounds?: string;
    workoutType: WorkoutType;
    workoutTypeDisplay?: string;
    exercisesToDoList: ExerciseViewModel[];
    children: WorkoutViewModel[];
    isInnerWorkout: boolean;

    
    constructor(params?: IWorkoutViewModel) {
      if (params == null) {
        return;
      }
      this.id = params.id;
      this.title = params.title;
      this.roundsCount = params.roundsCount;
      this.timeToWork = params.timeToWork;
      this.timeCap = params.timeCap;
      this.restBetweenExercises = params.restBetweenExercises;
      this.restBetweenRounds = params.restBetweenRounds;
      this.workoutType = params.workoutType;
      this.exercisesToDoList = params.exercisesToDoList;
      this.workoutTypeDisplay = params.workoutTypeDisplay;
      this.children = params.children;
      this.isInnerWorkout = params.isInnerWorkout;
    }

    public deserialize(jsonInput): WorkoutViewModel {
      if (jsonInput == null) {
        return null;
      }
      return new WorkoutViewModel({
        id: jsonInput.id,
        title: jsonInput.title,
        roundsCount: jsonInput.roundsCount,
        timeToWork: jsonInput.timeToWork,
        timeCap: jsonInput.timeCap,
        restBetweenExercises: jsonInput.restBetweenExercises,
        restBetweenRounds: jsonInput.restBetweenRounds,
        workoutType: jsonInput.workoutType,
        workoutTypeDisplay: jsonInput.workoutTypeDisplay,
        children: jsonInput.children.map(x => new WorkoutViewModel().deserialize(x)),
        exercisesToDoList: jsonInput.exercisesToDoList.map(x => new ExerciseViewModel().deserialize(x)),
        isInnerWorkout : false
      });
    }
  }
}

File: CrossfitDiary/CoreApp/CrossfitDiaryCore.Web/ViewModels/WorkoutViewModel.cs
using System;
using System.Collections.Generic;
using CrossfitDiaryCore.Model;
using Newtonsoft.Json;

namespace CrossfitDiaryCore.Web.ViewModels
{
    public class WorkoutViewModel
    {
        [JsonProperty("children")]
        public ICollection<WorkoutViewModel> Children { get; set; }

        [JsonProperty("exercisesToDoList")]
        public List<ExerciseViewModel> ExercisesToDoList { get; set; }

        [JsonProperty("id")]
        public int Id { get; set; }

        [JsonProperty("restBetweenExercises")]
        public string RestBetweenExercises { get; set; }

        [JsonProperty("restBetweenRounds")]
        public string RestBetweenRounds { get; set; }

        [JsonProperty("roundsCount")]
        public int? RoundsCount { get; set; }

        [JsonProperty("timeCap")]
        public string TimeCap { get; set; }

        [JsonProperty("timeToWork")]
        public string TimeToWork { get; set; }

        [JsonProperty("title")]
        public string Title { get; set; }

        [JsonProperty("workoutType")]
        public RoutineComplexType WorkoutType { get; set; }



        [JsonProperty("workoutTypeDisplay")]
        public string WorkoutTypeDisplay
        {
            get
            {
                string displayValue;
                switch (WorkoutType)
                {
                    case RoutineComplexType.ForTime:
                        displayValue = "For time";
                        break;
                    case RoutineComplexType.AMRAP:
                        displayValue = "AMRAP";
                        break;
                    case RoutineComplexType.EMOM:
                        displayValue = "EMOM";
                        break;
                    case RoutineComplexType.E2MOM:
                        displayValue = "E2MOM";
                        break;
                    case RoutineComplexType.NotForTime:
                        displayValue = "Not for time";
                        break;
                    case RoutineComplexType.ForTimeManyInners:
                        displayValue = "Complex for time";
                        break;
                    case RoutineComplexType.AMRAPN:
                        displayValue = "Complex amrap";
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }

                return displayValue;
            }
        }

        [JsonProperty("comment")]
        public string Comment { get; set; }

        [JsonProperty("resultsCount")]
        public int ResultsCount { get; set; }

        [JsonProperty("canSeeLeaderboard")]
        public bool CanSeeLeaderboard { get; set; }


        [JsonProperty("asNonBreakingSet")] 
        public bool AsNonBreakingSet { get; set; }
        
        
        [JsonProperty("findMaxWeight")] 
        public bool FindMaxWeight { get; set; }



    }
}